src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -456,5 +466,54 @@ internal void CopyTo(NpgsqlReadBuffer other)
         }
 
         #endregion
+
+
+        public IDictionary<string, long> RetrieveStatistics()
+        {
+            return _stats.RetrieveStatistics();
+        }
+
+        public void ResetStatistics()
+        {
+            _stats.ResetStatistics();
+        }
+    }
+
+    internal sealed class NpgsqlStatistics
+    {
+        private ConcurrentDictionary<string, long> _concurrentDictionary;
+
+        internal NpgsqlStatistics()
+        {
+            _concurrentDictionary = new ConcurrentDictionary<string, long>
+            {
+                ["BytesReceived"] = 0
+            };
+        }
+
+        public void AddBytesReceived(int bytesReceived)
+        {
+            _concurrentDictionary["BytesReceived"] += bytesReceived;
+        }
+
+        public IDictionary<string, long> RetrieveStatistics()
+        {
+            var dict = new Dictionary<string, long>();
+
+            foreach (var pair in _concurrentDictionary)
+            {
+                dict.Add(pair.Key, pair.Value);
+            }
+
+            return dict;
+        }
```

Comment:
Is there a reason not to pass `_concurrentDictionary` to the constructor?

```c#
var c = new ConcurrentDictionary<string, long>();
var d = new Dictionary<string, long>(c);
```

Reply:
Since the concurrent dictionary has a [thread-safe-but-non-snapshot](https://stackoverflow.com/a/12268686/7941718) enumerator, what happens if there's a write while you're enumerating? Would this end up returning mismatched statistics?